src/mongo_odm/model.hpp

```diff
@@ -0,0 +1,146 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cereal/cereal.hpp>
+
+#include <bsoncxx/oid.hpp>
+#include <mongo_odm/config/prelude.hpp>
+#include <mongo_odm/odm_collection.hpp>
+#include <mongocxx/collection.hpp>
+
+namespace mongo_odm {
+MONGO_ODM_INLINE_NAMESPACE_BEGIN
+
+template <typename T, typename IdType = bsoncxx::oid>
+class model {
+   private:
+// TODO: When XCode 8 is released, this can always be thread_local. Until then, the model class
+//       will not be thread-safe on OS X.
+#ifdef __APPLE__
+    static odm_collection<T> _coll;
+#else
+    static thread_local odm_collection<T> _coll;
+#endif
+
+   protected:
+    // TODO: if IdType is bsoncxx::oid, the derived class must initialize this until CXX-940 is
+    // resolved.
+    // if IdType is not bsoncxx::oid, the derived class is always responsible for initializing it.
+    IdType _id;
+
+   public:
+    /**
+     * Returns a copy of the underlying collection.
+     * @return The mongocxx::collection that this object wraps
+     */
+    static mongocxx::collection collection() {
```

Comment:
Hmm, if this a const ref, then a user can't call any of the underlying CRUD operations or drop(), which I was doing in one of the tests. I think we ideally want users to be able to do everything they need to do directly from the model, so I'm thinking that I'll keep it the way it is for now, but eventually once the model has everything a user would need, we can get rid of this method entirely. What do you think?